File: Programmers/[12929] 올바른 괄호의 갯수/ONE/12929.java
import java.util.*;

class Solution {
    private int answer = 0;
    public int solution(int n) {

        permutation(n, n, n, new ArrayList<>());

        return answer;
    }

    private void permutation(int n, int cntOpen, int cntClose, List<Integer> result) {
        if (cntOpen == 0 && cntClose == 0) {
            if (isRight(result))
                answer++;
            return;
        }
        if (cntOpen > 0) {
            result.add(1);
            permutation(n, cntOpen - 1, cntClose, result);
            result.remove(2 * n - cntClose - cntOpen);
        }
        if (cntClose > 0) {
            result.add(2);
            permutation(n, cntOpen, cntClose - 1, result);
            result.remove(2 * n - cntClose - cntOpen);
        }
    }

    private boolean isRight(List<Integer> result) {
        Stack<Integer> stack = new Stack<>();
        for (int num : result) {
            if (num == 1)
                stack.add(1);
            else {
                if (stack.isEmpty() || stack.peek() != 1)
                    return false;
                stack.pop();
            }
        }
        return true;
    }
}

File: Programmers/[12929] 올바른 괄호의 갯수/woong-jae/12929.js
function solution(n) {
  let result = 0;

  const stack = [];
  const helper = (used) => {
    if (used === n) {
      result++;
      return;
    }

    if (used < n) {
      stack.push("(");
      helper(used + 1);
      stack.pop();
    }
    const top = stack.length - 1;
    if (top >= 0 && stack[top] === "(") {
      stack.pop();
      helper(used);
      stack.push("(");
    }
  }
  helper(0);

  return result;
}